sectionfilter.c

```diff
@@ -270,54 +271,67 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
       Delete(i);
 }
 
+bool cSatipSectionFilterHandler::Send(void)
+{
+  // zero polling structures
+  memset(pollFdsM, 0, sizeof(pollFdsM));
+
+  // assemble all handlers to poll
+  for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+      if (filtersM[i] && filtersM[i]->Available() != 0) {
+         pollFdsM[i].fd = filtersM[i]->GetFd();
+         pollFdsM[i].events = POLLOUT;
+         }
+      }
+
+  // anyone ready for writing
+  if (poll(pollFdsM, eMaxSecFilterCount, eSecFilterSendTimeoutMs) <= 0)
+     return false;
+
+  // send data
+  for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+      if (pollFdsM[i].revents & POLLOUT)
+        filtersM[i]->Send();
+      }
+  return true;
+}
+
 void cSatipSectionFilterHandler::Action(void)
 {
   debug1("%s Entering [device %d]", __PRETTY_FUNCTION__, deviceIndexM);
-  bool processed = false;
   // Do the thread loop
+  uchar *p = NULL;
   while (Running()) {
+        int len = 0;
+        // Process all pending TS packets
+        while ((p  = ringBufferM->Get(len)) != NULL) {
+              if (p && (len >= TS_SIZE)) {
+                 if (*p != TS_SYNC_BYTE) {
+                    for (int i = 1; i < len; ++i) {
+                        if (p[i] == TS_SYNC_BYTE) {
+                           len = i;
+                           break;
+                           }
+                        }
+                    ringBufferM->Del(len);
+                    debug1("%s Skipped %d bytes to sync on TS packet [device %d]", __PRETTY_FUNCTION__, len, deviceIndexM);
+                    continue;
+                    }
+                    // Process TS packet through all filters
+                    mutexM.Lock();
+                    for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+                        if (filtersM[i])
+                           filtersM[i]->Process(p);
+                        }
+                    mutexM.Unlock();
+                    ringBufferM->Del(TS_SIZE);
+                    }
+                }
+
         // Send demuxed section packets through all filters
-        bool retry = false;
         mutexM.Lock();
-        for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
-            if (filtersM[i] && filtersM[i]->Send())
-               retry = true;
-            }
+        while (Send()) ;
```

Comment:
It might be better to include this "while" functionality inside the ```Send()``` method and rename it into e.g. ```SendAll()```.